File: scripts/fachascript_indentado.py
# fachascript_indentado.py
import sys
import os
import platform
import psutil
import time
from datetime import datetime

# Variables Base de FachaScript
variables = {
    'verdadero': True,
    'falso': False,
    'nulo': None,
    'cero': 0,
    'uno': 1,
    'pi': 3.1415926535,
    'e': 2.7182818284,
    'vacio': "",
    'usuario': os.getlogin(),
    'sistema': platform.system(),
    'ruta': os.getcwd(),
    'memoria_libre': psutil.virtual_memory().available,
    'cpu': platform.processor(),
    'procesos': [p.info for p in psutil.process_iter(['pid', 'name'])],
    'version': '0.1',
    'archivos': os.listdir(),
    'red_ip': None,  # Aquí debería estar el código para obtener la IP (dependiendo de la librería o plataforma)
    'modo_terminal': sys.stdout.isatty(),
    'contador': 0,
    'indice': 0,
    'limite': None,
    'iteracion_actual': 0,
    'estado_bucle': False,
    'entrada': '',
    'salida': '',
    'error': '',
    'resultado': None,
    'hoy': datetime.today().strftime('%Y-%m-%d'),
    'ahora': datetime.today().strftime('%H:%M:%S'),
    'zona_horaria': time.tzname[0],
    'tiempo_inicio': time.time(),
    'tiempo_ejecucion': 0,
    'facha_version': '0.1',
    'facha_sistema': platform.system(),
    'facha_errores': [],
    'facha_ruta_script': '',
    'facha_lenguaje': 'español',
    'facha_config': 'config.json',  # Puede ser un archivo JSON para configuración.
}

# Función para ejecutar el código de FachaScript
def ejecutar_fachascript(codigo):
    try:
        # Limpiamos posibles saltos de línea extra o comentarios
        codigo = codigo.strip()

        # Comandos básicos de FachaScript versión 0.1
        if codigo.startswith('imprimir'):
            # Comando 'imprimir' para mostrar salida en consola
            expresion = codigo[len('imprimir '):].strip()
            if expresion in variables:
                print(variables[expresion])
            else:
                print(expresion)
        
        elif codigo.startswith('crear'):
            # Comando 'crear' para crear variables
            variable = codigo[len('crear '):].strip()
            variables[variable] = None  # Crea una variable con valor None
            print(f"Variable '{variable}' creada.")
        
        elif codigo.startswith('asignar'):
            # Comando 'asignar' para asignar valor a una variable
            partes = codigo[len('asignar '):].split('=')
            if len(partes) == 2:
                variable = partes[0].strip()
                valor = partes[1].strip()
                if valor in variables:
                    variables[variable] = variables[valor]
                else:
                    variables[variable] = valor
                print(f"Variable '{variable}' asignada con valor '{valor}'.")
            else:
                print("Error: Asignación incorrecta.")
        
        elif codigo.startswith('sumar'):
            # Comando 'sumar' para hacer una suma
            partes = codigo[len('sumar '):].split('+')
            if len(partes) == 2:
                resultado = float(partes[0].strip()) + float(partes[1].strip())
                print(f"Resultado de la suma: {resultado}")
            else:
                print("Error: Operación de suma incorrecta.")
        
        elif codigo.startswith('restar'):
            # Comando 'restar' para hacer una resta
            partes = codigo[len('restar '):].split('-')
            if len(partes) == 2:
                resultado = float(partes[0].strip()) - float(partes[1].strip())
                print(f"Resultado de la resta: {resultado}")
            else:
                print("Error: Operación de resta incorrecta.")
        
        elif codigo.startswith('multiplicar'):
            # Comando 'multiplicar' para hacer una multiplicación
            partes = codigo[len('multiplicar '):].split('*')
            if len(partes) == 2:
                resultado = float(partes[0].strip()) * float(partes[1].strip())
                print(f"Resultado de la multiplicación: {resultado}")
            else:
                print("Error: Operación de multiplicación incorrecta.")
        
        elif codigo.startswith('dividir'):
            # Comando 'dividir' para hacer una división
            partes = codigo[len('dividir '):].split('/')
            if len(partes) == 2:
                divisor = float(partes[1].strip())
                if divisor != 0:
                    resultado = float(partes[0].strip()) / divisor
                    print(f"Resultado de la división: {resultado}")
                else:
                    print("Error: División por cero.")
            else:
                print("Error: Operación de división incorrecta.")
        
        elif codigo.startswith('si'):
            # Comando 'si' para hacer un condicional
            condicion = codigo[len('si '):].strip()
            if condicion.lower() == "verdadero":
                print("Condición es verdadera.")
            elif condicion.lower() == "falso":
                print("Condición es falsa.")
            else:
                print("Error: Condición inválida.")
        
        elif codigo.startswith('mientras'):
            # Comando 'mientras' para hacer un bucle while
            condicion = codigo[len('mientras '):].strip()
            if condicion.lower() == "verdadero":
                while True:
                    print("Bucle infinito.")
                    break  # Evitamos un bucle infinito para propósitos demostrativos.
            else:
                print("Error: Condición de 'mientras' no válida.")
        
        else:
            print(f"Error: Comando '{codigo}' no reconocido.")
    except Exception as e:
        print(f"Error al ejecutar el código: {e}")

# Función principal para leer y ejecutar el código desde un archivo
def ejecutar_archivo(archivo):
    try:
        with open(archivo, 'r') as f:
            codigo = f.read()
        # Ejecutar el código de FachaScript
        ejecutar_fachascript(codigo)
    except Exception as e:
        print(f"Error al leer el archivo: {e}")

# Código de ejecución
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Por favor, ingresa el nombre del archivo a ejecutar.")
    else:
        archivo = sys.argv[1]
        ejecutar_archivo(archivo)

File: scripts/fachascript_indentado.js
// fachascript_indentado.js
// Este archivo define los comandos y sus funcionalidades para ser utilizados en VS Code

const os = require('os');
const platform = require('os');
const psutil = require('psutil'); // npm install psutil
const time = require('time'); // npm install time
const { DateTime } = require('luxon'); // npm install luxon
const fs = require('fs');
const child_process = require('child_process');

// Variables base de FachaScript
const variables = {
    verdadero: true,
    falso: false,
    nulo: null,
    cero: 0,
    uno: 1,
    pi: 3.1415926535,
    e: 2.7182818284,
    vacio: "",
    usuario: os.userInfo().username,
    sistema: os.platform(),
    ruta: process.cwd(),
    version: 'FachaScript 0.1.3 Stable', // Versión fija
    facha_lenguage: "español"
};

// Comandos de FachaScript definidos en formato de funcionalidad

function imprimir(expresion) {
    if (expresion in variables) {
        return variables[expresion];
    } else {
        return expresion;
    }
}

function crear(variable) {
    variables[variable] = null;
    return `Variable '${variable}' creada.`;
}

function asignar(variable, valor) {
    if (valor in variables) {
        variables[variable] = variables[valor];
    } else {
        variables[variable] = valor;
    }
    return `Variable '${variable}' asignada con valor '${valor}'.`;
}

function sumar(valor1, valor2) {
    const resultado = parseFloat(valor1) + parseFloat(valor2);
    return `Resultado de la suma: ${resultado}`;
}

function restar(valor1, valor2) {
    const resultado = parseFloat(valor1) - parseFloat(valor2);
    return `Resultado de la resta: ${resultado}`;
}

function multiplicar(valor1, valor2) {
    const resultado = parseFloat(valor1) * parseFloat(valor2);
    return `Resultado de la multiplicación: ${resultado}`;
}

function dividir(valor1, valor2) {
    const divisor = parseFloat(valor2);
    if (divisor !== 0) {
        const resultado = parseFloat(valor1) / divisor;
        return `Resultado de la división: ${resultado}`;
    } else {
        return "Error: División por cero.";
    }
}

function si(condicion) {
    if (condicion.toLowerCase() === "verdadero") {
        return "Condición es verdadera.";
    } else if (condicion.toLowerCase() === "falso") {
        return "Error: Condición inválida.";
    } else {
        return "Error: Condición inválida.";
    }
}

function mientras(condicion) {
    if (condicion.toLowerCase() === "verdadero") {
        return "Bucle infinito simulado.";
    } else {
        return "Error: Condición no válida.";
    }
}

function comando_variable(nombre, valor) {
    variables[nombre] = valor;
    return `Variable '${nombre}' asignada con valor '${valor}'.`;
}

function comando_funcion(nombre, codigo) {
    // **Importante:** Esto solo almacena el código como una cadena. Para ejecutar funciones, necesitarías un intérprete.
    variables[nombre] = codigo;
    return `Función '${nombre}' definida (simulación).`;
}

function comando_ruta(nombre, ruta) {
    variables[nombre] = ruta;
    return `Variable '${nombre}' asignada con ruta '${ruta}'.`;
}

function usuario() {
    return os.userInfo().username;
}

function sistema() {
    return os.platform();
}

async function memoria_libre() {
    try {
        const mem = await psutil.virtualMemory();
        return mem.available;
    } catch (error) {
        return "Error al obtener la memoria libre. Asegúrate de haber instalado psutil correctamente.";
    }
}

function cpu() {
    try {
        const cpus = os.cpus();
        return `Modelo: ${cpus[0].model}, Núcleos: ${cpus.length}`;
    } catch (error) {
        return "Error al obtener información de la CPU.";
    }
}

async function lista_procesos() {
    try {
        const processes = await psutil.processList();
        const processList = processes.map(p => ({ pid: p.pid, name: p.name }));
        return JSON.stringify(processList); // Devuelve la lista como JSON
    } catch (error) {
        return "Error al obtener la lista de procesos. Asegúrate de haber instalado psutil correctamente.";
    }
}

async function archivos(ruta) {
    try {
        const files = fs.readdirSync(ruta); // Use synchronous method to be simpler
        return JSON.stringify(files); // Devuelve la lista como JSON
    } catch (error) {
        return `Error al leer el directorio: ${error}`;
    }
}

function red_ip() {
  try {
      // Ejecuta el comando ipconfig (Windows) o ifconfig (Linux/macOS)
      let command = os.platform() === 'win32' ? 'ipconfig' : 'ifconfig';
      const result = child_process.execSync(command).toString();
      return result
  } catch (error) {
      return `Error al obtener la IP: ${error}`;
  }
}

function modo_terminal() {
    // No hay forma confiable de detectar esto directamente en VS Code.  Podrías intentar usar un setting en la extensión
    return "No se puede determinar el modo terminal/GUI.";
}

function hoy() {
    return DateTime.now().toFormat('yyyy-MM-dd');
}

function ahora() {
    return DateTime.now().toFormat('HH:mm:ss');
}

function zona_horaria() {
    try {
        return DateTime.now().toFormat('ZZ');
    } catch (error) {
        return "Error al obtener la zona horaria.";
    }
}

function facha_sistema() {
    return os.platform();
}

function facha_ruta_script(ruta) {
    if (variables[ruta]) {
        return variables[ruta];
    } else {
        return ruta;
    }
}

function facha_lenguage() {
    return variables.facha_lenguage;
}

//Make it usable in typescript
module.exports = {
    imprimir,
    crear,
    asignar,
    sumar,
    restar,
    multiplicar,
    dividir,
    si,
    mientras,
    comando_variable,
    comando_funcion, // Added comando_funcion
    comando_ruta,
    usuario,
    sistema,
    memoria_libre,
    cpu,
    lista_procesos,
    archivos,
    red_ip,
    modo_terminal,
    hoy,
    ahora,
    zona_horaria,
    facha_sistema,
    facha_ruta_script,
    facha_lenguage
};